utils/logalloc.cc

```diff
@@ -360,9 +364,76 @@ static thread_local tracker& tracker_instance = get_tracker_instance();
 
 using clock = std::chrono::steady_clock;
 
+class background_reclaimer {
+    scheduling_group _sg;
+    noncopyable_function<void (size_t target)> _reclaim;
+    timer<lowres_clock> _adjust_shares_timer;
+    // If engaged, main loop is not running, set_value() to wake it.
+    promise<>* _main_loop_wait = nullptr;
+    future<> _done;
+    bool _stopping = false;
+    static constexpr size_t free_memory_threshold = 60'000'000;
+private:
+    bool have_work() const {
+#ifndef SEASTAR_DEFAULT_ALLOCATOR
+        return memory::stats().free_memory() < free_memory_threshold;
+#else
+        return false;
+#endif
+    }
+    void main_loop_wake() {
+        llogger.debug("background_reclaimer::main_loop_wake: waking {}", bool(_main_loop_wait));
+        if (_main_loop_wait) {
+            _main_loop_wait->set_value();
+            _main_loop_wait = nullptr;
+        }
+    }
+    future<> main_loop() {
+        llogger.debug("background_reclaimer::main_loop: entry");
+        while (true) {
+            while (!_stopping && !have_work()) {
+                promise<> wait;
+                _main_loop_wait = &wait;
+                llogger.trace("background_reclaimer::main_loop: sleep");
```

Comment:
should we perhaps adjust (i.e. reset) shares before going to sleep after we finished reclaiming? 

otherwise it may happen: 1) there's reclaiming work. 2) shares is set to a high value. 3) reclaiming work is done. 4) loop goes to sleep. 5) shares is still high even though there's no reclaiming work. it potentially confuses users observing scylla metrics. shares would only be readjusted next time there's reclaiming work.

Reply:
It's possible, but I don't want to adjust shares too often. It may confuse the scheduler.